File: tools/mql_transpiler/ast_nodes.py
"""MQL AST node types — pure data, zero logic, zero external dependencies.

These node types represent the parsed MQL source structure.  They are produced
by the CST→AST bridge (``ast_bridge.py``) and consumed by the codegen
(``ast_transpiler.py``).

No parser logic lives here.  No imports beyond the standard library.
"""

from __future__ import annotations


class ASTNode:
    """Base AST node."""


class SourceFile(ASTNode):
    def __init__(self, declarations=None):
        self.declarations = declarations


class VarDecl(ASTNode):
    def __init__(self, name="", var_type="", value=None, is_extern=False, is_input=False):
        self.name = name
        self.var_type = var_type
        self.value = value
        self.is_extern = is_extern
        self.is_input = is_input


class FuncDef(ASTNode):
    def __init__(self, name="", return_type="void", params=None, body=None):
        self.name = name
        self.return_type = return_type
        self.params = params
        self.body = body


class CompoundStmt(ASTNode):
    def __init__(self, statements=None):
        self.statements = statements


class ExpressionStmt(ASTNode):
    def __init__(self, expr=None):
        self.expr = expr


class IfStmt(ASTNode):
    def __init__(self, condition=None, then_branch=None, else_branch=None):
        self.condition = condition
        self.then_branch = then_branch
        self.else_branch = else_branch


class ForStmt(ASTNode):
    def __init__(self, init=None, condition=None, update=None, body=None):
        self.init = init
        self.condition = condition
        self.update = update
        self.body = body


class WhileStmt(ASTNode):
    def __init__(self, condition=None, body=None):
        self.condition = condition
        self.body = body


class SwitchStmt(ASTNode):
    def __init__(self, expr=None, cases=None, default=None):
        self.expr = expr
        self.cases = cases or []
        self.default = default


class ReturnStmt(ASTNode):
    def __init__(self, value=None):
        self.value = value


class Expression(ASTNode):
    """Base expression node."""


class TernaryExpr(Expression):
    """condition ? true_val : false_val"""
    def __init__(self, condition=None, true_val=None, false_val=None):
        self.condition = condition
        self.true_val = true_val
        self.false_val = false_val


class ArrayInitExpr(Expression):
    """{1.0, 2.0, 3.0} or array literal"""
    def __init__(self, elements=None):
        self.elements = elements or []


class BinaryOp(Expression):
    def __init__(self, left=None, op="", right=None):
        self.left = left
        self.op = op
        self.right = right


class UnaryOp(Expression):
    def __init__(self, op="", operand=None):
        self.op = op
        self.operand = operand


class CallExpr(Expression):
    def __init__(self, name="", args=None):
        self.name = name
        self.args = args


class SubscriptExpr(Expression):
    def __init__(self, name="", index=None):
        self.name = name
        self.index = index


class Identifier(Expression):
    def __init__(self, name=""):
        self.name = name


class NumberLiteral(Expression):
    def __init__(self, value=""):
        self.value = value


class StringLiteral(Expression):
    def __init__(self, value=""):
        self.value = value


class AssignmentExpr(Expression):
    def __init__(self, lhs="", rhs=None):
        self.lhs = lhs
        self.rhs = rhs

File: backend/tools/mql2go/ast_nodes.go
package mql2go

// AST node types for MQL — mirrors the Python ast_nodes.py.
// Produced by the tree-sitter CST→AST bridge.

// Node is the base interface for all AST nodes.
type Node interface{ node() }

// ── Top-level ─────────────────────────────────────────────────────

type SourceFile struct{ Declarations []Node }

// ── Declarations ──────────────────────────────────────────────────

type VarDecl struct {
	Name     string
	VarType  string
	Value    Node
	IsExtern bool
	IsInput  bool
}

type FuncDef struct {
	Name       string
	ReturnType string
	Params     []string
	Body       *CompoundStmt
}

type CompoundStmt struct{ Statements []Node }
type ExpressionStmt struct{ Expr Node }
type ReturnStmt struct{ Value Node }

// ── Control flow ──────────────────────────────────────────────────

type IfStmt struct {
	Condition  Node
	ThenBranch Node
	ElseBranch Node
}

type ForStmt struct {
	Init      Node
	Condition Node
	Update    Node
	Body      Node
}

type WhileStmt struct {
	Condition Node
	Body      Node
}

// ── Expressions ───────────────────────────────────────────────────

type BinaryOp struct {
	Left  Node
	Op    string
	Right Node
}

type UnaryOp struct {
	Op      string
	Operand Node
}

type CallExpr struct {
	Name string
	Args []Node
}

type Identifier struct{ Name string }
type NumberLiteral struct{ Value string }
type StringLiteral struct{ Value string }
type SubscriptExpr struct {
	Name  string
	Index Node
}
type AssignmentExpr struct {
	LHS string
	RHS Node
}

// Marker methods.
func (*SourceFile) node()      {}
func (*VarDecl) node()         {}
func (*FuncDef) node()         {}
func (*CompoundStmt) node()    {}
func (*ExpressionStmt) node()  {}
func (*ReturnStmt) node()      {}
func (*IfStmt) node()          {}
func (*ForStmt) node()         {}
func (*WhileStmt) node()       {}
func (*BinaryOp) node()        {}
func (*UnaryOp) node()         {}
func (*CallExpr) node()        {}
func (*Identifier) node()      {}
func (*NumberLiteral) node()   {}
func (*StringLiteral) node()   {}
func (*SubscriptExpr) node()   {}
func (*AssignmentExpr) node()  {}
